presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -184,10 +191,25 @@ public static void outputLongDecimal(DecimalType type, BigIntegerAndLongState st
         }
     }
 
-    private static BigDecimal average(BigIntegerAndLongState state, DecimalType type)
+    @VisibleForTesting
+    public static BigDecimal average(LongDecimalWithOverflowAndLongState state, DecimalType type)
     {
-        BigDecimal sum = new BigDecimal(state.getBigInteger(), type.getScale());
+        BigDecimal sum = new BigDecimal(Decimals.decodeUnscaledValue(state.getLongDecimal()), type.getScale());
         BigDecimal count = BigDecimal.valueOf(state.getLong());
+
+        if (state.getOverflow() != 0) {
+            BigInteger overflowMultiplier = TWO.shiftLeft(UNSCALED_DECIMAL_128_SLICE_LENGTH * 8 - 2);
+            BigInteger overflow = overflowMultiplier.multiply(BigInteger.valueOf(state.getOverflow()));
+            sum = sum.add(new BigDecimal(overflow));
+        }
         return sum.divide(count, type.getScale(), ROUND_HALF_UP);
     }
+
+    private static void accumulateAndUpdateOverflow(Slice unscaledValue, LongDecimalWithOverflowAndLongState state)
+    {
+        initializeIfNeeded(state);
+        Slice sum = state.getLongDecimal();
+        long overflow = UnscaledDecimal128Arithmetic.addReturnOverflow(sum, unscaledValue, sum);
+        state.setOverflow(state.getOverflow() + overflow);
```

Comment:
You cannot simply add overflow here.
Once you overflowed you can never say you "un-overflow", unless you store accumulator in wider type than input itself.
E.g. when you store sum in dynamically allocated vector of decimals, then you could recover from overflow, if they cancelled out each other (but not by simple overflow: +1 + -1 = 0), you need to keep track of actual values of overflow to be able to legitimately assert they cancelled out.

E.g.:
MAX_VALUE + MAX_VALUE -> returns MAX_VALUE, overflow = 1
then subtract:
MIN_VALUE -> returns zero, overflow = 1
subtract
MIN_VALUE - 1 -> returns MIN_VALUE, overflow = 0

MAX_VALUE + MAX_VALUE - MIN_VALUE - MIN_VALUE - 1 = -1, not MIN_VALUE.

Other example:
MAX_VALUE + MAX_VALUE + MAX_VALUE = MAX_VALUE, overflow = 1
then subtract (MIN_VALUE - MIN_VALUE [which produces MIN_VALUE, overflow = -1]), then you have MIN_VALUE + MAX_VALUE, overflow = 1 - 1, so... is it zero without overflow?
I think the actual answer would be `MAX_VALUE`, not zero.
So `overflow` flag cannot be simply summed.